remotes/docker/fetcher.go

```diff
@@ -103,20 +103,16 @@ func (r dockerFetcher) open(ctx context.Context, u, mediatype string, offset int
 		// really distinguish between a 206 and a 200. In the case of 200, we
 		// can discard the bytes, hiding the seek behavior from the
 		// implementation.
-
 		defer resp.Body.Close()
+
 		if resp.StatusCode == http.StatusNotFound {
 			return nil, errors.Wrapf(errdefs.ErrNotFound, "content at %v not found", u)
 		}
-		body, err := ioutil.ReadAll(resp.Body)
-		if err == nil {
-			dockerErr := errcode.Errors{}
-			err := json.Unmarshal(body, &dockerErr)
-			if err == nil && dockerErr.Len() > 0 {
-				return nil, errors.Errorf("unexpected status code %v: %s - Server message: %s", u, resp.Status, dockerErr.Error())
-			}
+		var registryErr errcode.Errors
+		if err := json.NewDecoder(resp.Body).Decode(&registryErr); err != nil || registryErr.Len() < 1 {
+			return nil, errors.Errorf("unexpected status code %v: %v", u, resp.Status)
 		}
-		return nil, errors.Errorf("unexpected status code %v: %v", u, resp.Status)
+		return nil, errors.Wrapf(registryErr, "unexpected status code %v: %v", u, resp.Status)
```

Comment:
The original PR was submitted to expose servers (registries) providing specific details to help users understand why a specific action failed (e.g. billing, quota, etc.); I think using `Wrapf` with the specific error as the `cause` means it is up to the `Fetch` caller whether that ends up in the output (e.g. kubectl describe list of image pull failures, for one example). I guess maybe @bainsy88 can test and validate whether that's an issue. I don't know how to generate the right response to test it :)

Reply:
fixed